pkg/gatewayserver/gatewayserver_test.go

```diff
@@ -664,9 +657,30 @@ func TestGatewayServer(t *testing.T) {
 					t.Run(tc.Name, func(t *testing.T) {
 						a := assertions.New(t)
 
+						testCtx := test.ContextWithT(test.Context(), t)
+						testCtx, cancel := context.WithTimeout(testCtx, 5*timeout)
```

Comment:
What is the benefit of this?

Now if operation 1 takes longer than `timeout`, it passes on to operation 2 which may not have enough time to finish, causing the wrong part of the test to fatal.

Reply:
The `time.After(timeout)` was used in the `select` statement in order to prevent deadlocks, right? AFAIK there's no predefined time interval, within which an operation has to finish.
Given this, this approach achieves exactly the same goal as the old `time.After(timeout)`, however not it's done on per-testcase basis, not per-operation.
1. The `ctx.Done()` can be used not only by the test cases, but also by the functions, taking `ctx` (not relevant here, but is often the case in other tests)
2. If an operation 1 took a bit longer than `timeout`, the test case should not fail IMO, it should continue execution, since the `timeout` value is just an arbitrary value we chose to avoid deadlocks.

Please note that the context timeout is `5*timeout`, which seemed to approximate the worst-case timeout ceiling used in previous approach.

Reply:
To address case you describe exactly: if operation 1 took longer than `timeout`, but succeeded, then the `timeout` value is too low

Reply:
> The `time.After(timeout)` was used in the `select` statement in order to prevent deadlocks, right?

No, it is used for testing timeouts of steps.

>AFAIK there's no predefined time interval, within which an operation has to finish.

It's not about the time interval, it is about giving each step a certain time to finish.

Again, what is the benefit of a timeout context? I get the context cancellation part for the test, but simply not the timeout.

> To address case you describe exactly: if operation 1 took longer than `timeout`, but succeeded, then the `timeout` value is too low

The problem is that if you have a timeout of 100 and you have 5 operations, that there's a timeout after 500, right? So if operation 1 takes 490, operation 2 is very unlikely to finish in the remaining 10, so it gives a "operation 2 timeout" error, while that is not the correct fatal.